composeApp/src/commonMain/kotlin/data/network/errorhandling/ApiErrorInterceptor.kt

```diff
@@ -0,0 +1,31 @@
+package data.network.errorhandling
+
+import di.ApiErrorValidator
+import io.ktor.client.statement.HttpResponse
+import io.ktor.http.HttpStatusCode
+
+class ApiErrorInterceptor : ApiErrorValidator {
+    override fun validateResponse(response: HttpResponse) {
+        when (val statusCode = response.status) {
+            HttpStatusCode.Forbidden -> throw ApiErrorException(statusCode, "Access denied")
+            HttpStatusCode.BadRequest -> throw ApiErrorException(statusCode, "Bad request")
+            HttpStatusCode.Unauthorized -> throw ApiErrorException(statusCode, "Unauthorized")
+            HttpStatusCode.NotFound -> throw ApiErrorException(statusCode, "Resource not found")
+            HttpStatusCode.InternalServerError -> throw ApiErrorException(
+                statusCode,
+                "Internal server error",
+            )
+            HttpStatusCode.ServiceUnavailable -> throw ApiErrorException(
+                statusCode,
+                "Service unavailable",
+            )
+            HttpStatusCode.RequestTimeout -> throw ApiErrorException(
+                statusCode,
+                "The request timed out",
+            )
+            else -> {
+                throw ApiErrorException(statusCode, "Unexpected status code: ${statusCode.value}")
+            }
+        }
```

Comment:
This when statement doesn't seem that necessary to me, all it does is checking the code and putting it right back to another class.  The message can be an extension function based on the status code and it should not be hardcoded (if it is meant for UI that is)

Reply:
I think what we really should do here is to try and convert non 200 responses and their response bodies into domain error exception. I.e something like that
```data class ApiException(
    val code: Int,
    val displayableMessage: String,
    val loggableMessage: String
) : IOException(displayableMessage)
```